File: JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

// (c) Copyright [2022-2023] Open Text.
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.SQLTransientConnectionException;
import java.sql.SQLInvalidAuthorizationSpecException;
import java.util.Map;
import java.util.HashMap;
import java.util.Properties;
import java.io.IOException;
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.net.URLEncoder;
import java.net.HttpURLConnection;
import java.net.URL;
import com.google.gson.JsonParser;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.vertica.jdbc.VerticaConnection;

public class OAuthSampleApp {
	private static final String OAUTH_ACCESS_TOKEN_VAR_STRING = "OAUTH_SAMPLE_ACCESS_TOKEN";
	private static final String OAUTH_REFRESH_TOKEN_VAR_STRING = "OAUTH_SAMPLE_REFRESH_TOKEN";

	private static Properties prop;
	private static Map<String, String> csProp = new HashMap<String, String>();
	private static Connection vConnection;

	// main function, Call starts from here
	public static void main(String[] args) {
		try {
			loadProperties();
			setUpDbForOAuth();
			ensureAccessToken();
			connectToDatabase();
		} catch (SQLException e) {
			e.printStackTrace();
		} catch (Exception unreportedEx) {
			unreportedEx.printStackTrace();
		} finally {
			tearDown();
		}
		System.exit(0);
	}

	// admin must log in and setup the user for OAuth
	// steps can also be followed in vsql
	private static void setUpDbForOAuth() throws Exception{
		String connectionString = getConnectionString();
		String user = prop.getProperty("User");
		String clientId = prop.getProperty("ClientId");
		String clientSecret = prop.getProperty("ClientSecret");
		String discoveryUrl = prop.getProperty("DiscoveryUrl");
		String queries[]= {
				"DROP USER IF EXISTS " + user + " CASCADE;",
				"DROP AUTHENTICATION IF EXISTS v_oauth CASCADE;",
				"CREATE AUTHENTICATION v_oauth METHOD 'oauth' LOCAL;",
				"ALTER AUTHENTICATION v_oauth SET client_id = '" + clientId + "';",
				"ALTER AUTHENTICATION v_oauth SET client_secret = '" + clientSecret + "';",
				"ALTER AUTHENTICATION v_oauth SET discovery_url = '" + discoveryUrl + "';",
				"CREATE USER " + user + ";",
				"GRANT AUTHENTICATION v_oauth TO " + user + ";",
				"GRANT ALL ON SCHEMA PUBLIC TO " + user + ";"
				};
		vConnection = DriverManager.getConnection(connectionString);
		Statement st = vConnection.createStatement();
		
		for (String query: queries) {
			st.execute(query);
		}
		st.close();
	}

	// removes the oauth user and closes the admin connection
	private static void tearDown() {
		try {
			Statement st = vConnection.createStatement();
			String user = prop.getProperty("User");
			st.executeUpdate("DROP USER IF EXISTS " + user + " CASCADE");
			st.executeUpdate("DROP AUTHENTICATION IF EXISTS v_oauth CASCADE");
			vConnection.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	// If access token is Invalid/Expired, Get new tokens using password/refresh grant
	private static void ensureAccessToken() throws Exception {
		String accessToken = System.getenv(OAUTH_ACCESS_TOKEN_VAR_STRING);
		String refreshToken = System.getenv(OAUTH_REFRESH_TOKEN_VAR_STRING);
		try {
			if (null == accessToken || accessToken.isEmpty()) {
				if (null == refreshToken || refreshToken.isEmpty()) {
					// Obtain first access token and refresh Tokens
					System.out.println("Access token is invalid/expired, trying to do refresh");
					getTokensByPasswordGrant();
				} else {
					try {
						System.out.println("Attempting to use given refresh token.");
						getTokensByRefreshGrant();
					} catch (Exception e) {
						System.out.println("Initial refresh token has expired. Getting new access and refresh tokens.");
						getTokensByPasswordGrant();
					}
				}
			} else {
				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, accessToken);
			}
		} catch (Exception e) {
			throw e;
		}
	}

	// Test connection using access token and test database query result
	private static void connectToDatabase() throws SQLException {
		int connAttemptCount = 0;
		while (connAttemptCount <= 1) {
			try {
				String accessToken = System.getProperty(OAUTH_ACCESS_TOKEN_VAR_STRING);
				if (null == accessToken || accessToken.isEmpty()) {
					throw new Exception("Access Token is not available.");
				} else {
					System.out.println("Attempting to connect with OAuth access token");
					Connection conn = getConnection(accessToken);
					ResultSet rs = executeQuery(conn);
					printResults(rs);
					System.out.println("Query Executed. Exiting.");
					break;
				}
			} catch (Exception ex) {
				if (connAttemptCount > 0) {
					break;
				}
				try {
					getTokensByRefreshGrant();
				} catch (Exception e1) {
					try {
						System.out.println("Refresh token is invalid/Expired, Getting new tokens");
						getTokensByPasswordGrant();
					} catch (Exception e2) {
						e2.printStackTrace();
					}
				}
				++connAttemptCount;
			}
		}
	}

	// password grant logs into the IDP using credentials in app.config
	public static void getTokensByPasswordGrant() throws Exception {
		Map<String, String> formData = new HashMap<String, String>();
		formData.put("grant_type", "password");
		formData.put("client_id", prop.getProperty("ClientId"));
		formData.put("client_secret", prop.getProperty("ClientSecret"));
		formData.put("username", prop.getProperty("User"));
		formData.put("password", prop.getProperty("Password"));
		getAndSetTokens(formData);
	}

	// refresh grant uses the refresh token to get the new access and refresh token
	public static void getTokensByRefreshGrant() throws Exception {
		Map<String, String> formData = new HashMap<String, String>();
		formData.put("grant_type", "refresh_token");
		formData.put("client_id", prop.getProperty("ClientId"));
		formData.put("client_secret", prop.getProperty("ClientSecret"));
		formData.put("refresh_token", System.getProperty(OAUTH_REFRESH_TOKEN_VAR_STRING));
		getAndSetTokens(formData);
	}

	// get and set tokens from IDP
	private static void getAndSetTokens(Map<String, String> formData) throws Exception {
		try {
			String postOpts = getEncodedParamsString(formData);
			byte[] postData = postOpts.getBytes("UTF-8");
			int postDataLength = postData.length;
			URL url = new URL(prop.getProperty("TokenUrl"));
			HttpURLConnection connection = (HttpURLConnection) url.openConnection();
			try {
				connection.setDoOutput(true);
				connection.setUseCaches(false);
				connection.setRequestMethod("POST");
				connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
				connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
				connection.setRequestProperty("Accept", "application/json");
				connection.getOutputStream().write(postData);
				BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
				ByteArrayOutputStream buf = new ByteArrayOutputStream();
				readResult(in, buf);
				String res = buf.toString("UTF-8");
				JsonElement jElement = JsonParser.parseString(res);
				JsonObject jObject = jElement.getAsJsonObject();
				// set Tokens as System Properties - new values to access_token and
				// refresh_token
				String accessToken = jObject.has("access_token") ? jObject.get("access_token").getAsString() : null;
				String refreshToken = jObject.has("refresh_token") ? jObject.get("refresh_token").getAsString() : null;
				if (null == accessToken) {
					throw new Exception(
							"Access/refresh token is null, Please verify the config.properties for proper inputs.");
				}
				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, accessToken);
				if (null != refreshToken && !refreshToken.isEmpty()) {
					System.setProperty(OAUTH_REFRESH_TOKEN_VAR_STRING, refreshToken);
				}
			} catch (UnsupportedEncodingException uee) {
				uee.printStackTrace();
			} catch (Exception e) {
				String res = "";
				try {
					BufferedInputStream in = new BufferedInputStream(connection.getErrorStream());
					ByteArrayOutputStream buf = new ByteArrayOutputStream();
					readResult(in, buf);
					res = buf.toString("UTF-8");
				} catch (Exception ex) {
					// Skip when this happens, but print the error.
					ex.printStackTrace();
				}
				throw e;
			} finally {
				connection.disconnect();
			}
		} catch (IOException unreportedex) {
			unreportedex.printStackTrace();
		}
	}

	// Get jdbc connection string from provided configuration
	private static void getcsProp(String args) {
		String[] entries = args.split(";");
		for (String entry : entries) {
			String[] pair = entry.split("=");
			csProp.put(pair[0], pair[1]);
		}
	}
	
	// get connection string for admin connection to setup OAuth Record
	private static String getConnectionString() {
		return "jdbc:vertica://" + csProp.get("Host") + ":" + csProp.get("Port") + "/" + csProp.get("Database")
				+ "?user=" + csProp.get("User") + "&password=" + csProp.get("Password");
	}

	// Get the parameters from the connection String
	private static String getParam(String paramName) {
		return csProp.get(paramName);
	}

	// Connect to Database using access Token
	private static Connection getConnection(String accessToken) throws SQLException {
		Properties jdbcOptions = new Properties();
		jdbcOptions.put("oauthaccesstoken", accessToken);
		return DriverManager.getConnection(
				"jdbc:vertica://" + getParam("Host") + ":" + getParam("Port") + "/" + getParam("Database"),
				jdbcOptions);
	}

	// execute Simple query on database connection
	private static ResultSet executeQuery(Connection conn) throws SQLException {
		Statement stmt = conn.createStatement();
		String query = "SELECT user_id, user_name FROM USERS ORDER BY user_id";
		System.out.println("Executing query:" + query);
		return stmt.executeQuery(query);
	}

	private static void printResults(ResultSet rs) throws SQLException {
		int rowIdx = 1;
		while (rs.next()) {
			System.out.println(rowIdx + ". " + rs.getString(1).trim() + " " + rs.getString(2).trim());
			rowIdx++;
		}
	}

	// Get encoded URL from parameters
	private static String getEncodedParamsString(Map<String, String> params) throws UnsupportedEncodingException {
		StringBuilder result = new StringBuilder();
		for (Map.Entry<String, String> entry : params.entrySet()) {
			result.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8.name()));
			result.append("=");
			result.append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8.name()));
			result.append("&");
		}
		result.setLength(result.length() - 1);
		return result.toString();
	}

	// read result from Buffered input stream
	private static ByteArrayOutputStream readResult(BufferedInputStream in, ByteArrayOutputStream buf) {
		try {
			for (int result = in.read(); result != -1; result = in.read()) {
				buf.write((byte) result);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return buf;
	}

	// load the configuration properties
	public static void loadProperties() {
		prop = new Properties();
		try (InputStream input = new FileInputStream("src/main/java/config.properties")) {
			// load the properties file
			prop.load(input);
			// Get the connectionString parameters from properties prop
			getcsProp(prop.getProperty("ConnectionString"));
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}
}

File: ADO.NET/OAuthSampleApp/OAuthSampleApp.cs
using Newtonsoft.Json.Linq;
using System.Configuration;
using System.Runtime.InteropServices;
using Vertica.Data.VerticaClient;

internal class OAuthSampleApp
{
    private static VerticaConnectionStringBuilder connectionStringBuilder;
    private static VerticaConnection vConnection;
    private const string OAUTH_ACCESS_TOKEN_VAR_STRING = "OAUTH_SAMPLE_ACCESS_TOKEN";
    private const string OAUTH_REFRESH_TOKEN_VAR_STRING = "OAUTH_SAMPLE_REFRESH_TOKEN";

    static async Task Main(string[] args)
    {
        connectionStringBuilder = new VerticaConnectionStringBuilder();
        connectionStringBuilder.ConnectionString = ConfigurationManager.ConnectionStrings["ConnectionString"].ConnectionString;

        await SetUpDbForOAuth();
        await EnsureAccessToken();
        await ConnectToDatabase();
        await TearDown();
    }

    // admin must log in and setup the user for OAuth
    // steps can also be followed in vsql
    static async Task SetUpDbForOAuth()
    {
        vConnection = new VerticaConnection(connectionStringBuilder.ConnectionString);
        vConnection.Open();

        string USER = getSetting("User");
        string CLIENT_ID = getSetting("ClientId");
        string CLIENT_SECRET = getSetting("ClientSecret");
        string INTROSPECT_URL = getSetting("TokenUrl") + "introspect";

        VerticaCommand cmd = new VerticaCommand("", vConnection);
        string[] sqlCommands =
        {
            "DROP USER IF EXISTS " + USER + " CASCADE",
            "DROP AUTHENTICATION IF EXISTS adooauth CASCADE",
            "CREATE USER " + USER,
            "CREATE AUTHENTICATION adooauth METHOD 'oauth'" + GetAuthHost(),
            "GRANT AUTHENTICATION adooauth TO " + USER,
            "GRANT ALL ON SCHEMA PUBLIC TO " + USER,
            "ALTER AUTHENTICATION adooauth SET client_id = '" + CLIENT_ID + "'",
            "ALTER AUTHENTICATION adooauth SET client_secret = '" + CLIENT_SECRET + "'",
            "ALTER AUTHENTICATION adooauth SET introspect_url = '" + INTROSPECT_URL + "'",
        };

        // Execute each SQL command in the array
        foreach (string commandText in sqlCommands)
        {
            cmd.CommandText = commandText;
            cmd.ExecuteNonQuery();
        }
    }

    // removes the oauth user and closes the admin connection
    public static async Task TearDown()
    {
        string USER = getSetting("User");
        VerticaCommand cmd = new VerticaCommand("DROP USER IF EXISTS " + USER + " CASCADE", vConnection);
        cmd.ExecuteNonQuery();
        cmd.CommandText = "DROP AUTHENTICATION IF EXISTS adooauth CASCADE";
        cmd.ExecuteNonQuery();

        vConnection.Close();
    }

    // if no access token is set, obtains and sets first access and refresh tokens
    // uses the password grant flow
    static async Task EnsureAccessToken()
    {
        string refreshToken = Environment.GetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
        string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);

        if (string.IsNullOrEmpty(accessToken))
        {
            if (string.IsNullOrEmpty(refreshToken))
            {
                // Obtain first access token
                Console.WriteLine("Access token not found. Obtaining first access token from IDP.");
                await GetTokensByPasswordGrant();
            }
            else
            {
                try
                {
                    // attempt to use given refresh token
                    Console.WriteLine("Attempting to use given refresh token.");
                    await GetTokensByRefreshGrant();
                }
                catch (HttpRequestException hre)
                {
                    Console.WriteLine(hre.Message);
                    Console.WriteLine("Initial refresh token has expired. Getting new access and refresh tokens.");
                    await GetTokensByPasswordGrant();
                }
            }
        } else
        {
            Console.WriteLine("Access token found in environment variable.");
        }

        SetAccessToken();
    }

    // the access token must be set in the connection string
    static void SetAccessToken()
    {
        string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
        connectionStringBuilder["OAuthAccessToken"] = accessToken;
    }

    // OAuth user attempts to connect with the access token
    // If it fails, we try to get a new access token with the refresh token
    // If the refresh token is expired, we get new access and refresh tokens
    public static async Task ConnectToDatabase()
    {
        // ADO.NET requires a user in the connection string
        // It is ignored by the driver, and only the access token is used
        // The user can be set to any value (except the admin) and it will connect
        connectionStringBuilder.User = getSetting("User");
        int connAttemptCount = 0;

        while (connAttemptCount <= 1)
        {
            try
            {
                using (VerticaConnection conn = new VerticaConnection(connectionStringBuilder.ConnectionString))
                {
                    Console.WriteLine("Attempting to connect with connection string: " + ResultSetPrinter.printableConnectionString(connectionStringBuilder.ConnectionString));
                    conn.Open();

                    using (VerticaCommand command = conn.CreateCommand())
                    {
                        Console.WriteLine("Connection Successful");

                        // change the query in app.config or make the command text a string literal
                        command.CommandText = getSetting("Query");
                        using (VerticaDataReader reader = command.ExecuteReader())
                        {
                            Console.WriteLine("Executing Query: " + command.CommandText);
                            ResultSetPrinter printer = new ResultSetPrinter(reader);
                            printer.printResults();
                            Console.WriteLine("Query Executed. Exiting.");

                            break;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                if (connAttemptCount > 0) { break; }
                try
                {
                    Console.WriteLine("Access token invalid or expired. Attempting token refresh.");
                    await GetTokensByRefreshGrant();
                }
                catch (HttpRequestException hre)
                {
                    Console.WriteLine(hre.Message);
                    Console.WriteLine("Refresh token invalid or expired. Attempting to get new access and refresh tokens");
                    await GetTokensByPasswordGrant();
                }

                SetAccessToken();
                connAttemptCount++;
            }
        }
    }

    // password grant logs into the IDP using credentials in app.config
    public static async Task GetTokensByPasswordGrant()
    {
        Dictionary<string, string> formData = new Dictionary<string, string>
        {
            {"grant_type",   "password"},
            {"client_id",     getSetting("ClientId")},
            {"client_secret", getSetting("ClientSecret")},
            {"username",      getSetting("User")},
            {"password",      getSetting("Password")}
        };

        await GetAndSetTokens(formData);
    }

    // refresh grant uses the refresh token to get a new access and refresh token
    public static async Task GetTokensByRefreshGrant()
    {
        Dictionary<string, string> formData = new Dictionary<string, string>
        {
            {"grant_type",   "refresh_token"},
            {"client_id",     getSetting("ClientId")},
            {"client_secret", getSetting("ClientSecret")},
            {"refresh_token", Environment.GetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, EnvironmentVariableTarget.User)}
        };

        await GetAndSetTokens(formData);
    }

    // uses the token url from the config to make an http request
    // will always always set new access and refresh tokens, even if currently valid
    private static async Task GetAndSetTokens(Dictionary<string, string> formData)
    {
        HttpClient httpClient = new HttpClient();

        try
        {
            var content = new FormUrlEncodedContent(formData);

            string tokenUrl = getSetting("TokenUrl");
            var response = await httpClient.PostAsync(tokenUrl, content);

            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadAsStringAsync();
            JObject jObject = JObject.Parse(result);

            // Set the access token as an environment variable
            Environment.SetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, jObject["access_token"].ToString(), EnvironmentVariableTarget.User);

            // Set the refresh token as an environment variable if IDP returns one
            if (jObject["refresh_token"] != null)
            {
                Environment.SetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, jObject["refresh_token"].ToString(), EnvironmentVariableTarget.User);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error getting refresh and/or access tokens: " + ex.Message);
            throw;
        }
    }

    static string getSetting(String setting)
    {
        return ConfigurationManager.AppSettings[setting] ?? "";
    }

    // default hosts depending on OS
    static string GetAuthHost()
    {
        return RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? "LOCAL" : "HOST '0.0.0.0/0'";
    }
}

